flink-libraries/flink-table/src/main/scala/org/apache/flink/table/formats/TableFormatFactoryService.scala

```diff
@@ -0,0 +1,245 @@
+      val supportedKeys = normalizeSupportedProperties(factory)
+      throw new NoMatchingTableFormatException(
+        s"""
+          |The matching factory '${factory.getClass.getName}' doesn't support '${lastKey.get}'.
+          |
+          |Supported properties of this factory are:
+          |${supportedKeys.sorted.mkString("\n")}""".stripMargin,
+        factoryClass,
+        foundFactories,
+        properties)
+    } else if (supportedFactories.isEmpty) {
+      throw new NoMatchingTableFormatException(
+        s"No factory supports all properties.",
+        factoryClass,
+        foundFactories,
+        properties)
+    } else if (supportedFactories.length > 1) {
+      throw new AmbiguousTableFormatException(
+        supportedFactories,
+        factoryClass,
+        foundFactories,
+        properties)
+    }
+
+    supportedFactories.head.asInstanceOf[T]
+  }
+
+  private def findMatchingContext(
+      properties: Map[String, String],
+      classLoader: ClassLoader)
+    : (Seq[TableFormatFactory[_]], Seq[TableFormatFactory[_]]) = {
+
+    val foundFactories = mutable.ArrayBuffer[TableFormatFactory[_]]()
+    val matchingFactories = mutable.ArrayBuffer[TableFormatFactory[_]]()
+
+    try {
+      val iter = if (classLoader == null) {
+        defaultLoader.iterator()
+      } else {
+        val customLoader = ServiceLoader.load(classOf[TableFormatFactory[_]], classLoader)
+        customLoader.iterator()
+      }
+      while (iter.hasNext) {
+        val factory = iter.next()
+        foundFactories += factory
+
+        val requestedContext = normalizeContext(factory)
+
+        val plainContext = mutable.Map[String, String]()
+        plainContext ++= requestedContext
+        // we remove the version for now until we have the first backwards compatibility case
+        // with the version we can provide mappings in case the format changes
+        plainContext.remove(FORMAT_PROPERTY_VERSION)
+
+        // check if required context is met
+        if (plainContext.forall(e => properties.contains(e._1) && properties(e._1) == e._2)) {
+          matchingFactories += factory
+        }
+      }
+    } catch {
+      case e: ServiceConfigurationError =>
+        LOG.error("Could not load service provider for table format factories.", e)
+        throw new TableException("Could not load service provider for table format factories.", e)
+    }
+
+    (foundFactories, matchingFactories)
+  }
+
+  private def normalizeContext(factory: TableFormatFactory[_]): Map[String, String] = {
+    val requiredContextJava = factory.requiredContext()
+    if (requiredContextJava != null) {
+      requiredContextJava.asScala.map(e => (e._1.toLowerCase, e._2)).toMap
+    } else {
+      Map[String, String]()
+    }
+  }
+
+  private def normalizeSupportedProperties(factory: TableFormatFactory[_]): Seq[String] = {
+    val supportedPropertiesJava = factory.supportedProperties()
+    if (supportedPropertiesJava != null) {
```

Comment:
How can `null` happen? Either change interface to `Optional<List<...>>` or add `checkNotNull(...)` here.

Reply:
This is just a safety check. Because factories might be implemented by users. According to the documentation this should not be null. But it does hurt to check for null on API borders.